### ISSUE TITLE
Safari (Mac/iPad/iPhone): hyperHTML.wire() with static template does not guarantee the same generated element

### ISSUE THREAD

**Issue by ikabirov** [WebReflection/hyperHTML#338]
Tagged template function is not guaranteed to receive the same instance of a template array on Safari, even if it is used with the same template literal. It causes hyperHTML.wire() to generate a different instance of an HTMLElement.

Due to this issue it is not possible to make a component without adding wrapping element.
Here is the simplest example reproducing this issue.
https://codepen.io/anon/pen/QYwoRy 
Run it on Safari and wait for some time. On our devices it usually takes less than minute to see an alert.

It seems that Safari needs the _templateLiteral function fix similar to fix for Firefox<55.

**Comment by WebReflection** [comment]
I have removed all the unnecessary stuff to reproduce the issue, and I am going to file a bug to WebKit.

This is the most annoying bug ever kind, 'cause it's impossible to feature detect.

However, are you sure you want/need to retain any node reference at all? The fact nodes are preserved in hyperhtml shouldn't translate into "_feel free to reference these and retain these around_" because that's an implementation/performance detail you shouldn't care about.

In which scenario would this be an issue? I am just trying to understand if it's worth it to penalize Safari perf due random glitches on their GC.

**Comment by ikabirov** [comment]
> In which scenario would this be an issue?

I use it for creating component without top level wrapper. E.g now:
```
// button with internal state
import Button from 'button.js'

document.body.appendChild(Button({
  text: 'Save'
})
```

If result element can be changed, we must create wrapper for every component, and result DOM will be something like:
```
<body>
  <div>
    <button>Save</button>
  </div>
</body>
```

Assumption that element should not be changed allow us to remove unnecessary wrapper and receive something like:
```
<body>
  <button>Save</button>
</body>
```

**Comment by WebReflection** [comment]
could you bind a document fragment instead, as unique holder of the button?

```js
const wm = new WeakMap;
const Button = identity => wm.get(identity) || set(identity);
function set(identity) {
  const fragment = document.createDocumentFragment();
  const render = hyperHTML.bind(fragment);
  wm.set(identity, render);
  return render`<button>${identity.text}</button>`;
}
```

or something similar ?

**Comment by WebReflection** [comment]
sorry, I know I haven't really provided a solution, just a ugly work around.

Let's see what Safari/WebKit folks say about their bug, so I'll figure out what I should do here.

**Comment by ikabirov** [comment]
@WebReflection thank you! I'll try your work around.

**Comment by ikabirov** [comment]
> could you bind a document fragment instead, as unique holder of the button?

it doesn't work, cause after `document.body.appendChild(fragment)` all button will be removed from fragment and moved to body. So after Button re-render it will be moved back to fragment and removed from body

**Comment by WebReflection** [comment]
it looks like they've closed the bug as duplicated, 'cause lit-html opened the same issue already.

Of course if it's Google complaining they listen and fix so ... good lit-html has the exact same issue.

I don't think I'll do anything for the time being, but I'll keep this open in case I change my mind.

**Comment by ikabirov** [comment]
May be add option to force use/ignore _templateLiteral function fix? 

For safari ES6 it would be great to always turn it on (ES6 used in our debug environment)
For production we use Closure Compiler with transpilation to ES5. it would be great to always turn it off (now property raw is enumerable, so we have unused overhead)

What about hyperHTML.forceTemplateLiteralFix?

**Comment by WebReflection** [comment]
the template literal fix is a third part dependency, nothing affected by random properties in this or that library.

https://github.com/ungap/template-literal#template-literal

**Comment by WebReflection** [comment]
So ... this would force Safari to the slow path, which I believe is the simplest/best option for the time being.

https://github.com/ungap/template-literal/pull/3

The ugly sniff/hack will be removed once they fix it in a future Safari version.

Your production code will still use the slow path, but:

  1. better safe than sorry
  2. iOS devices are already the fastest one out there, it won't hurt
  3. this might put more pressure in WebKit/Apple folks to fix this properly, since they care about being slowed down

Please try to use that patch in your env and tell me if everything is fine, thanks.

**Comment by ikabirov** [comment]
thank you

**Comment by WebReflection** [comment]
@ikabirov don't thank me yet, it's not planned to go out until you confirm it works as expected 😅

**Comment by ikabirov** [comment]
@WebReflection it seems everything is fine

**Comment by WebReflection** [comment]
So I've published it as `0.2.0` 'cause even if it's a patch it changes slightly the behavior of the utility.

Whenever I'll find time to update all related projects, I will, meanwhile you can force-use that version.
